compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -517,13 +519,14 @@ impl<'tcx> GeneratorSubsts<'tcx> {
 
     #[inline]
     pub fn upvar_tys(self) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
```

Comment:
Nit: I think this function merits a comment too. Something like

> Returns an iterator over the upvar types for this closure/generator. This should only be invoked after type check is complete; before that, the upvar types may still be being inferred or may be an inference variable. To access the upvar types before type check is known to have completed, use `tupled_upvar_tys`. 

That said, this will still ICE in the event of a `Ty::Err`, right? I'm not sure if that's a great idea, as it means that every use of this function has to be guarded by a call to `tupled_upvars_ty`. Why not make this function return a `Result` as well (or perhaps an empty vector)?